File: Java/src/main/java/de/keineInsektenImEssen/textdetector/service/UnwantedIngredientsProvider.java
package de.keineInsektenImEssen.textdetector.service;

import com.google.common.collect.ImmutableSet;
import de.keineInsektenImEssen.textdetector.model.Pattern;
import de.keineInsektenImEssen.textdetector.model.Synonym;

import java.util.List;
import java.util.Set;

import static java.util.regex.Pattern.CASE_INSENSITIVE;
import static java.util.regex.Pattern.compile;

public class UnwantedIngredientsProvider {

    // FK-TODO: zu jedem der folgenden unerwünschten Zutaten in openfoodfacts ein Beispielprodukt suchen und die JSON-Antwort auf unerwünschte Zutaten korrekt parsen (siehe https://de.openfoodfacts.org/cgi/search.pl?action=display&tagtype_0=ingredients&tag_contains_0=contains&tag_0=KARMIN&sort_by=unique_scans_n&page_size=20)
    public static final Set<Synonym> unwantedIngredients =
            ImmutableSet.of(
                    new Synonym(
                            "E120",
                            List.of(
                                    new Pattern("E120", compile("E\\s*120")),
                                    new Pattern("Schildlaus", compile("Schildlaus", CASE_INSENSITIVE)),
                                    new Pattern("Karmin", compile("Karmin", CASE_INSENSITIVE)))),
                    new Synonym(
                            "E904",
                            List.of(
                                    new Pattern("E904", compile("E\\s*904")),
                                    new Pattern("Schellack", compile("Schellack", CASE_INSENSITIVE)))),
                    new Synonym(
                            "Tenebrio molitor",
                            List.of(
                                    new Pattern("Tenebrio molitor", compile("Tenebrio\\s+molitor", CASE_INSENSITIVE)),
                                    new Pattern("Mehlkäfer", compile("Mehlkäfer", CASE_INSENSITIVE)),
                                    new Pattern("Mehlwurm", compile("Mehlwurm|Mehlwürmer", CASE_INSENSITIVE)))),
                    new Synonym(
                            "Locusta migratoria",
                            List.of(
                                    new Pattern("Locusta migratoria", compile("Locusta\\s+migratoria", CASE_INSENSITIVE)),
                                    new Pattern("Wanderheuschrecke", compile("Wanderheuschrecke", CASE_INSENSITIVE)))),
                    new Synonym(
                            "Acheta domesticus",
                            List.of(
                                    new Pattern("Acheta domesticus", compile("Acheta\\s+domesticus", CASE_INSENSITIVE)),
                                    new Pattern("Hausgrille", compile("Hausgrille", CASE_INSENSITIVE)))),
                    new Synonym(
                            "Alphitobius diaperinus",
                            List.of(
                                    new Pattern("Alphitobius diaperinus", compile("Alphitobius\\s+diaperinus", CASE_INSENSITIVE)),
                                    new Pattern("Buffalowurm", compile("Buffalowurm|Buffalowürmer", CASE_INSENSITIVE)),
                                    new Pattern("Getreideschimmelkäfer", compile("Getreideschimmelkäfer", CASE_INSENSITIVE)))));
}

File: InsectsInFoodChecker/UnwantedIngredientsProvider.js
class UnwantedIngredientsProvider {

    static synonyms =
        [
            new Synonym(
                'E120',
                [
                    new Pattern('E120', /E\s*120/g),
                    new Pattern('Schildlaus', /Schildlaus/gi),
                    new Pattern('Karmin', /Karmin/gi)
                ],
                new URL("https://en.wikipedia.org/wiki/Carmine")),
            new Synonym(
                'E904',
                [
                    new Pattern('E904', /E\s*904/g),
                    new Pattern('Schellack', /Schellack/gi)
                ],
                new URL("https://en.wikipedia.org/wiki/Shellac")),
            new Synonym(
                'Tenebrio molitor',
                [
                    new Pattern('Tenebrio molitor', /Tenebrio\s+molitor/gi),
                    new Pattern('Mehlkäfer', /Mehlkäfer/gi),
                    new Pattern('Mehlwurm', /Mehlwurm|Mehlwürmer/gi)
                ],
                new URL("https://en.wikipedia.org/wiki/Mealworm")),
            new Synonym(
                'Locusta migratoria',
                [
                    new Pattern('Locusta migratoria', /Locusta\s+migratoria/gi),
                    new Pattern('Wanderheuschrecke', /Wanderheuschrecke/gi)
                ],
                new URL("https://en.wikipedia.org/wiki/Migratory_locust")),
            new Synonym(
                'Acheta domesticus',
                [
                    new Pattern('Acheta domesticus', /Acheta\s+domesticus/gi),
                    new Pattern('Hausgrille', /Hausgrille/g)
                ],
                new URL("https://en.wikipedia.org/wiki/House_cricket")),
            new Synonym(
                'Alphitobius diaperinus',
                [
                    new Pattern('Alphitobius diaperinus', /Alphitobius\s+diaperinus/gi),
                    new Pattern('Buffalowurm', /Buffalowurm|Buffalowürmer/gi),
                    new Pattern('Getreideschimmelkäfer', /Getreideschimmelkäfer/gi)
                ],
                new URL("https://en.wikipedia.org/wiki/Alphitobius_diaperinus"))
        ];
}